src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/RSACng.EncryptDecrypt.cs

```diff
@@ -0,0 +1,106 @@
+        // api's invoke this common helper with the "transform" parameter determining whether encryption or decryption is done.
+        //
+        private byte[] EncryptOrDecrypt(byte[] data, RSAEncryptionPadding padding, EncryptOrDecryptAction encryptOrDecrypt)
+        {
+            if (data == null)
+                throw new ArgumentNullException("data");
+
+            if (padding == null)
+                throw new ArgumentNullException("padding");
+
+            unsafe
+            {
+                SafeNCryptKeyHandle keyHandle = Key.Handle;
+                switch (padding.Mode)
+                {
+                    case RSAEncryptionPaddingMode.Pkcs1:
+                        return EncryptOrDecrypt(keyHandle, data, AsymmetricPaddingMode.NCRYPT_PAD_PKCS1_FLAG, null, encryptOrDecrypt);
+
+                    case RSAEncryptionPaddingMode.Oaep:
+                        {
+                            using (SafeUnicodeStringHandle safeHashAlgorithmName = new SafeUnicodeStringHandle(padding.OaepHashAlgorithm.Name))
+                            {
+                                BCRYPT_OAEP_PADDING_INFO paddingInfo = new BCRYPT_OAEP_PADDING_INFO()
+                                {
+                                    pszAlgId = safeHashAlgorithmName.DangerousGetHandle(),
+
+                                    // It would nice to put randomized data here but RSAEncryptionPadding does not at this point provide support for this.
+                                    pbLabel = IntPtr.Zero,
+                                    cbLabel = 0, 
+                                };
+                                return EncryptOrDecrypt(keyHandle, data, AsymmetricPaddingMode.NCRYPT_PAD_OAEP_FLAG, &paddingInfo, encryptOrDecrypt);
+                            }
+                        }
+
+                    default:
+                        throw new CryptographicException(SR.Cryptography_UnsupportedPaddingMode);
+                }
+            }
+        }
+
+        //
+        // Now that the padding mode and information have been marshaled to their native counterparts, perform the encryption or decryption.
+        //
+        private static unsafe byte[] EncryptOrDecrypt(SafeNCryptKeyHandle key, byte[] input, AsymmetricPaddingMode paddingMode, void* paddingInfo, EncryptOrDecryptAction encryptOrDecrypt)
+        {
+            int numBytesNeeded;
+            ErrorCode errorCode = encryptOrDecrypt(key, input, input.Length, paddingInfo, null, 0, out numBytesNeeded, paddingMode);
+            if (errorCode != ErrorCode.ERROR_SUCCESS)
+                throw errorCode.ToCryptographicException();
+
+            byte[] output = new byte[numBytesNeeded];
+            errorCode = encryptOrDecrypt(key, input, input.Length, paddingInfo, output, numBytesNeeded, out numBytesNeeded, paddingMode);
+            if (errorCode != ErrorCode.ERROR_SUCCESS)
```

Comment:
Is it ever possible at this point for numBytesNeeded to be less than output.Length?  i.e. could the first call tell us a certain number of bytes were needed, but actually use fewer on the next call?  I assume not, but maybe an assert would be helpful.